### PR TITLE
AB#51 Implement Recording Engine

### PR DESCRIPTION
**Pull request by StuckiSimon**
Add base implementation for user script recording engine.

The scheduling engine is primitive and will be extended as part of other stories. Handling priorities of user intents is not part of this PR.

The code within the MoveIntent is currently just as a proposal. This PR does not handle position conflicts.

Furthermore the current script execution context is not secure as discussed in the previous meetings this won't be handled in the current sprint.

### CODE REVIEW

File: game-backend/src/game-engine/intent/MoveIntent.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  execute(state: State, userContext: UserContext): State {
+    const unit = state.objectMap[userContext.unitId];
+
+    const position = {
+      ...unit.position,
+    };
+    switch (this.direction) {
+      case Direction.Up:
+        position.x--;
```
Comment (sir-marc) [inline game-backend/src/game-engine/intent/MoveIntent.ts]: Consider using the x parameter as the vertical axis.

File: game-backend/src/game-engine/intent/MoveIntent.test.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            },
+          },
+        },
+        objectIds: ['unit1'],
+      },
+      {
+        unitId: 'unit1',
+      }
+    );
+    expect(newState.objectMap.unit1.position.x).toBe(1);
```
Comment (sir-marc) [inline game-backend/src/game-engine/intent/MoveIntent.test.ts]: TDD at its best :D

**Review by sir-marc — CHANGES_REQUESTED** [review]
Other than that it looks great!

File: game-backend/src/game-engine/intent/MoveIntent.ts (hunk 1 above)
Comment (StuckiSimon) [reply game-backend/src/game-engine/intent/MoveIntent.ts]: Let's handle these kind of discussions as part of the actual implementation. Atm it's really only about the engine :v:

**Review by sir-marc — APPROVED** [review]

**Pull request merged**